Support timing function for animation. Fix #12

phimage opened the pull request:
- Timing function enum added
  - Add it to `Animatable` objects and `AnimationConfiguration`
  - Use this enum when possible instead of `CAMediaTimingFunction`

TODO
- [x] add more type in enum (see `extension CAMediaTimingFunction`)
- [x] add more type in enum  see `CAMediaTimingFunction(` contructors in code
- [x] add custom type with control points in enum ?
- [x] manage spring animation or not, animation default value
- [x] Add an example
- [x] Update CHANGELOG
- [x] Documentation

phimage committed:
Add timing function enum. Add it to Animatable object

lastMove commented on IBAnimatable/TimingFunctionType.swift:
```diff
@@ -0,0 +1,210 @@
+      self = .default
+
+    // customs
+    case "easeInSine":
```
It's a little bit confusing but `extractNameAndParams` returns only lowercased string (I will add it to the function documentation) so the `case` have to be lowercase too. (It's this way so case is ignored via InterfaceBuilder

phimage committed:
Add more TimingFunctionType

phimage commented:
added more types using sourcery to generate switch case from static variables in `CAMediaTimingFunction`

 stencil template 
```
{% for type in types.all %}

{% for variable in type.variables %}
case {{ variable.name }}
{% endfor %}

{% for variable in type.variables %}
case .{{ variable.name }}:
return .{{ variable.name }}
{% endfor %}

{% for variable in type.variables %}
case "{{ variable.name|lowercase }}":
self = .{{ variable.name }}
{% endfor %}

{% endfor %}
```

lastMove commented on IBAnimatable/Animatable.swift:
```diff
@@ -44,9 +44,14 @@ public protocol Animatable: class {
 }
 
-public extension Animatable where Self: UIView {
+public extension Animatable {
```
Nice one

phimage replied:
Forgot to remove this change for this PR
(Trying to make `UIBarItem` also `Animatable`, I see that I can remove that)

lastMove commented on IBAnimatable/Animatable.swift:
```diff
@@ -327,12 +334,12 @@ fileprivate extension Animatable where Self: UIView {
       squashY.values = [1, 0.5, 1, 0.5, 1]
       squashY.keyTimes = [0, 0.2, 0.4, 0.6, 0.8, 1]
-      squashY.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
+      squashY.timingFunctionType = .easeInOut
```
Should allow to customize the timingFunction for squash animation

lastMove commented on IBAnimatable/Animatable.swift:
```diff
@@ -327,12 +334,12 @@ fileprivate extension Animatable where Self: UIView {
       squashX.values = [1, 1.5 * configuration.force, 0.5, 1.5 * configuration.force, 1]
       squashX.keyTimes = [0, 0.2, 0.4, 0.6, 0.8, 1]
-      squashX.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
+      squashX.timingFunctionType = .easeInOut
```
Should allow to customize the timingFunction for squash animation

lastMove reviewed (dismissed):
Great work ! 
It's almost OK for me.
It could be useful to add the possibility to choose the timingFunction in the "Animations" Section of  the IBAnimatable App.

phimage committed:
Manage default timing function type by animation if `none` selected
Use timing function type in squash animation
Add `custom` and `spring` timing function type
Use easeOutExpo in orbit activity indicator

phimage commented:
default value managed using operator `??` to replace `none`with a default one at the right of the operator

I think I made a fix in `ActivityIndicatorAnimationOrbit` ring2OpacityAnimation
same could be done in ring1OpacityAnimation


For the demo
maybe as @lastMove say in the "Animations" Section 
or a new one with
- two translation animations
- two or one timing function type menu selectors
this allow to compare two timing functions (could compare always with linear if only one menu selector)

lastMove commented:
> a new one with
two translation animations
two or one timing function type menu selectors
this allow to compare two timing functions (could compare always with linear if only one menu selector)

 Nice Idea !
It's a great first step.
The inconvenient of this solution is that we can see the effect of the timing function only on Translations.

phimage commented:
Maybe I will do the two case
1 - in Animations add a button to display a new view controller to select the timing function
![screen shot 2017-05-07 at 11 54 28](https://cloud.githubusercontent.com/assets/8875768/25779565/f852c5d0-331b-11e7-8444-e2f527143005.png)
I am working on it today

2 - add a new entry with two elements, one animation (maybe later an animation type selector in a new view controller)
a picker for timing function

phimage committed:
In demo App allow to select animations timing functions
using picker and a view, which display the bezier curve of selected function

phimage commented:
![mediatimingfunction](https://cloud.githubusercontent.com/assets/8875768/25781103/f85ab952-3334-11e7-892d-c2f6ff010807.gif)
added in "Animations" a view to select the Timing Function as illustrated in the previous message

EDIT: added two falling stars to see effect of Timing Function
Also moveBy and moveTo could now use cubic animation with timing function instead of spring animation if timing function defined

Need some help for the Documentations, where to document it?
- APIs.md there is the new Animatable protocol attribute

phimage committed:
Make moveTo and moveBy compatible with TimingFunctionType
If none, use the spring animation
Add two falling stars in demo app to show TimingFunction effect

tbaranes commented:
@phimage Sorry for the delay, do you still need help for the documentation?
